### PR TITLE
Gemma4_26B_A4B_NvFp4 hf checkpoint convert to gguf format fixes

### PR DESCRIPTION
**Pull request by ynankani**
## Overview
Gemma4_26B_A4B_NvFp4 hf checkpoint convert to gguf format fixes. This PR fixes the following:
1) Excluded weight_scale, weight_scale_2, and input_scale from the existing + ".weight" rename for .experts. tensors. The original rename was causing issue with NVFP4 scale tensor names (e.g. experts.0.down_proj.weight_scale_2 => experts.0.down_proj.weight_scale_2.weight), breaking the NVFP4 lookup at _generate_nvfp4_tensors
2) Added FFN_GATE_EXP, FFN_UP_EXP, alongside the existing FFN_GATE_UP_EXP in the GEMMA4 tensor allow-list. Originally only fused FFN_GATE_UP_EXP was allowed. HF NVFP4 checkpoints store gate/up/down as separate per-expert tensors, so the converter couldn't map them especially for NvFP4 . Other option was to re-quantize if want to fuse gate and up proj.
3) made ffn_gate_up_exps TENSOR_NOT_REQUIRED/Optional, added fallback creation of separate ffn_gate_exps and ffn_up_exps if the fused tensor is absent
4) Conditional plumbing in build_moe_ffn so that it passes either fused or separate tensors
5) Pre-folds each layer's router.per_expert_scale into the corresponding expert's down_proj.weight_scale_2 at conversion time, then pop()s router.per_expert_scale from model_tensors so the existing modify_tensors mapping doesn't fire for NVFP4 conversions which was causing "Duplicated tensor name 'blk.0.ffn_down_exps.scale' error" as both per_expert_scale and weight_scale_2 were using same slot.
## Additional information

Tested with https://huggingface.co/nvidia/Gemma-4-26B-A4B-NVFP4

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: Yes, Took AI assistance for debugging and refactoring.

### CODE REVIEW

File: src/models/gemma4.cpp [hunk 1]
```diff
@@ -286,8 +292,8 @@ llama_model_gemma4::graph::graph(const llama_model & model, const llm_graph_para
 
             cur_moe = build_moe_ffn(cur_moe,
                     nullptr, // gate_inp
-                    nullptr, // up_exps
-                    nullptr, // gate_exps
+                    model.layers[il].ffn_gate_up_exps ? nullptr : model.layers[il].ffn_up_exps, // up_exps
+                    model.layers[il].ffn_gate_up_exps ? nullptr : model.layers[il].ffn_gate_exps, // gate_exps
```
Comment (CISC) [inline src/models/gemma4.cpp]:
```suggestion
                    model.layers[il].ffn_up_exps,
                    model.layers[il].ffn_gate_exps,
```

File: src/models/gemma4.cpp [hunk 2]
```diff
@@ -296,8 +302,8 @@ llama_model_gemma4::graph::graph(const llama_model & model, const llm_graph_para
                     LLAMA_EXPERT_GATING_FUNC_TYPE_SOFTMAX,
                     il, logits,
                     model.layers[il].ffn_gate_up_exps,
-                    nullptr, // up_exps_s
-                    nullptr, // gate_exps_s
+                    model.layers[il].ffn_gate_up_exps ? nullptr : model.layers[il].ffn_up_exps_s, // up_exps_s
+                    model.layers[il].ffn_gate_up_exps ? nullptr : model.layers[il].ffn_gate_exps_s, // gate_exps_s
```
Comment (CISC) [inline src/models/gemma4.cpp]:
```suggestion
                    model.layers[il].ffn_up_exps_s,
                    model.layers[il].ffn_gate_exps_s,
```

File: convert_hf_to_gguf.py [hunk 3]
```diff
@@ -7988,13 +7988,34 @@ def generate_extra_tensors(self) -> Iterable[tuple[str, Tensor]]:
...
+
     @classmethod
     def filter_tensors(cls, item: tuple[str, Callable[[], Tensor]]) -> tuple[str, Callable[[], Tensor]] | None:
         name, gen = item
 
         if name.endswith("per_dim_scale") or name.endswith("layer_scalar"):
             name = name + ".weight"
-        if ".experts." in name and not name.endswith(".weight"):
+        if ".experts." in name and not name.endswith(".weight") \
+                and not name.endswith(("weight_scale", "weight_scale_2", "input_scale")):
```
Comment (CISC) [inline convert_hf_to_gguf.py]:
```suggestion
        if ".experts." in name and not name.endswith((".weight", ".weight_scale", ".weight_scale_2", ".input_scale")):
```

**Review by CISC — APPROVED** [review]

File: convert_hf_to_gguf.py [hunk 4]
```diff
@@ -7988,13 +7988,33 @@ def generate_extra_tensors(self) -> Iterable[tuple[str, Tensor]]:
...
+                continue
+            bid = bid_match.group(1)
+            r = self.model_tensors.pop(name)
+            for e in range(n_experts):
+                w2 = f"model.language_model.layers.{bid}.experts.{e}.down_proj.weight_scale_2"
+                if w2 in self.model_tensors:
+                    s = self.model_tensors[w2]
+                    self.model_tensors[w2] = lambda s=s, r=r, i=e: s() * r()[i]
+        super()._generate_nvfp4_tensors()
+
```
Comment (gaugarg-nv) [inline convert_hf_to_gguf.py:8014]: I think this can have correctness issues. What if a model doesn't contain down proj expert in NVFP4? This change silently drops `router.per_expert_scale` without checking if all matrix are actually NVFP4.

Can you also add a sanaity check to make sure there is a at least one matching `w2` before popping `router.per_expert_scale`?

Comment (gaugarg-nv) [reply convert_hf_to_gguf.py:8014]: @ynankani Can you check this part of the code: https://github.com/ynankani/llama.cpp/blob/9678f8812c5c0b9e1e9fd7f22724a56a95c834cb/convert_hf_to_gguf.py#L539

It seems to remove `language_model` string from all tensors. That means no `w2` will match and there is no folding of multipliers. Only `per_expert_scale` is being dropped.

Comment (CISC) [reply convert_hf_to_gguf.py:8014]: Nice catch, this was perhaps tested before I introduced this change?

**Comment by catlilface** [comment]
Thank you for your contribution. I've successfully built and launched Gemma4 26b NVFP4 but unfortunately faced performance issues with CPU offloading. On my setup token generation drops sub 10tps even if only a few layers offloaded, when Q8 quant generates at more than 60tps.
Since I haven't seen anything that could lead to these issues in your pr I suppose it's related to NVFP4 implementation itself.

File: convert_hf_to_gguf.py (hunk 4 above)
Comment (ynankani) [reply convert_hf_to_gguf.py:8014]: Yes, this was tested without the latest change in filter_tensor.

**Review by gaugarg-nv — APPROVED** [review]

**Comment by CISC** [comment]
BTW, it doesn't look like we support fused gate/up experts in `_generate_nvfp4_tensors`, but since Gemma4 got split, maybe this is not a thing (at least with ModelOpt)?

**Comment by CISC** [comment]
@ynankani Looks like GitHub UI snuck in `\r\n` again, can you normalize to `\n`?

**Comment by ynankani** [comment]
> BTW, it doesn't look like we support fused gate/up experts in `_generate_nvfp4_tensors`, but since Gemma4 got split, maybe this is not a thing (at least with ModelOpt)?

I see ModelOpt treating gate and up experts as separate for quantization and export [Link](https://github.com/NVIDIA/Model-Optimizer/blob/e2d29c869b97aae5cfacd5be4db079951a43e2ac/modelopt/torch/export/layer_utils.py#L1019).
So it should be fine as separate gate and up are exported.

Also I see logic for calibration sync for gate and up proj[Link](https://github.com/NVIDIA/Model-Optimizer/blob/e2d29c869b97aae5cfacd5be4db079951a43e2ac/modelopt/torch/export/layer_utils.py#L1197 ) [Link2](https://github.com/NVIDIA/Model-Optimizer/blob/e2d29c869b97aae5cfacd5be4db079951a43e2ac/modelopt/torch/export/moe_utils.py#L26)
So the serving engine can fuse the gate and up at inference time.

**Comment by ynankani** [comment]
> @ynankani Looks like GitHub UI snuck in `\r\n` again, can you normalize to `\n`?

Is it possible in .gitattributes to set eol=lf for python file?

**Comment by CISC** [comment]
> > @ynankani Looks like GitHub UI snuck in `\r\n` again, can you normalize to `\n`?
> 
> Is it possible in .gitattributes to set eol=lf for python file?

I'm not sure what the reason is, but it seems totally random, so I doubt there's anything we can do.

**Review by CISC — APPROVED** [review]
